rust-runtime/aws-smithy-runtime-api/src/client/http/request.rs

```diff
@@ -0,0 +1,476 @@
+    /// See [HeaderMap::get]
+    pub fn get(&self, key: impl AsRef<str>) -> Option<&str> {
+        self.headers.get(key.as_ref()).map(|v| v.as_ref())
+    }
+
+    /// Returns the total number of **values** stored in the map
+    pub fn len(&self) -> usize {
+        self.headers.len()
+    }
+
+    /// Returns true if this header is present
+    pub fn contains_key(&self, key: &str) -> bool {
+        self.headers.contains_key(key)
+    }
+
+    /// Insert a value into the headers structure.
+    ///
+    /// This will *replace* any existing value for this key. Returns the previous associated value if any.
+    ///
+    /// # Panics
+    /// If the key or value are not valid ascii, this function will panic.
+    pub fn insert(
+        &mut self,
+        key: impl AsHeaderComponent,
+        value: impl AsHeaderComponent,
+    ) -> Option<String> {
+        self.try_insert(key, value)
+            .expect("HeaderName or HeaderValue was invalid")
+    }
+
+    /// Insert a value into the headers structure.
+    ///
+    /// This will *replace* any existing value for this key. Returns the previous associated value if any.
+    ///
+    /// If the key or value are not valid ascii, an error is returned
+    pub fn try_insert(
+        &mut self,
+        key: impl AsHeaderComponent,
+        value: impl AsHeaderComponent,
+    ) -> Result<Option<String>, HttpError> {
+        let key = header_name(key.into_maybe_static()?)?;
+        let value = header_value(value.into_maybe_static()?)?;
+        // try to convert it into our more restrictive header value
+        let value: HeaderValue = value.try_into().map_err(HttpError::invalid_header_value)?;
+        Ok(self
+            .headers
+            .insert(key, value)
+            .map(|old_value| old_value.into()))
+    }
+
+    /// Appends a value to a given key
+    ///
+    /// If the key or value are NOT valid ascii, an error is returned
+    pub fn try_append(
+        &mut self,
+        key: impl AsHeaderComponent,
+        value: impl AsHeaderComponent,
+    ) -> Result<bool, HttpError> {
+        let key = header_name(key.into_maybe_static()?)?;
+        let value = header_value(value.into_maybe_static()?)?;
+        Ok(self.headers.append(key, value))
+    }
+
+    /// Appends a value to a given key
+    ///
+    /// # Panics
+    /// If the key or value are NOT valid ascii, this function will panic
+    pub fn append(&mut self, key: impl AsHeaderComponent, value: impl AsHeaderComponent) -> bool {
+        self.try_append(key, value)
+            .expect("HeaderName or HeaderValue was invalid")
+    }
+}
+
+use sealed::AsHeaderComponent;
+
+mod sealed {
+    use super::*;
+    /// Trait defining things that may be converted into a header component (name or value)
+    pub trait AsHeaderComponent {
+        fn into_maybe_static(self) -> Result<MaybeStatic, HttpError>;
+        fn already_header_name(self) -> Result<http0::HeaderName, Self>
```

Comment:
```suggestion
        fn already_header_name_http_03x(self) -> Result<http0::HeaderName, Self>
```
?

Or, since we're always dealing with concrete objects instead of dyn traits, I think we could even make this generic over the return type so the caller can request whatever concrete type they want. Although, implementing that may be quite gross...

I guess this isn't too much of a concerns since this is a private trait though.

Reply:
changed the name to `try_as_http03_header_name`